File: src/recall/beat_relay.py
"""LAN fallback for the mic heartbeat: the Mac accepts a beat and passes it to the
fleet (#888).

WHY THIS EXISTS. The beat's reachability requirement used to be STRICTER than
recording's, which made the check lie about working phones:

    audio   phone -> Mac,  192.168.1.81:9999   (LAN)
    beat    phone -> Isis, 10.100.0.2:8000     (VPN)

So a phone sitting at home with its tunnel off streamed every sample correctly and
still read `silent` after 3 h and `dead` after 12 h. Both false — the mic was
working. That is the error #837 was built to remove (a quiet room and a dead app
reading alike), reintroduced one layer up. Measured twice on 2026-08-14, when the
iPhone's WireGuard was off and `devicectl` kept reaching it over the LAN the whole
time, which is exactly what made it look like an app fault.

WHY IT IS NOT THE INGEST PORT. The obvious home for this is the port the phones
already talk to, but the ingest server CLOSES its listener while capture is
paused — and a pause is precisely when the heartbeat is the only signal there is.
A beat receiver has to be independent of the capture lifecycle, so it is its own
tiny server that runs whatever capture is doing.

WHY PORT 8000. The same port the fleet API answers on, so the apps need no second
URL shape: the fallback is the identical request with `host` swapped for
`controlHost`. It is a different machine, so nothing collides.

⚠ ISIS REMAINS THE SINGLE SOURCE OF TRUTH. This forwards; it does not store. A
local beat store the collector had to merge with the fleet's would let two places
disagree about which mics are alive, which is worse than the bug being fixed.

⚠ UNAUTHENTICATED, like the endpoint it forwards to — a mic app has never held a
credential, and a beat that could 401 would report a credential mistake as dead
hardware. But this one listens on the LAN rather than behind the VPN, so it
forwards an ALLOWLIST of known fields and stamps `viaLan` itself: a caller here
can lie about its own state (as any phone always could) but cannot inject fields
into the fleet store or deny that it came the back way.

Stdlib only, deliberately: this runs on the Mac beside the capture agent, and the
import-surface checks keep that process free of the API's dependencies.
"""

from __future__ import annotations

import json
import logging
import urllib.error
import urllib.request
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from typing import Any, Final, override

_log = logging.getLogger("recall.beat_relay")

#: Same port as the fleet API, on a different machine — see the module docstring.
DEFAULT_RELAY_PORT: Final = 8000

#: Where a relayed beat goes. Isis over the Mac's one-way VPN (the Mac may
#: initiate into the VPN by design), which is why this direction works at all.
DEFAULT_FLEET_URL: Final = "http://10.100.0.2:8000"

BEAT_PATH: Final = "/api/devices/heartbeat"

#: Bounded because the caller is unauthenticated: a beat is a few hundred bytes.
MAX_BODY_BYTES: Final = 4096
MAX_DEVICE_LEN: Final = 64
FORWARD_TIMEOUT_S: Final = 8

#: The fleet answers 200; this relay answers 204 (it stores nothing, so it has no
#: body to return). Both are "the beat landed", and insisting on 200 would have made
#: every relayed beat read as a failure to the phone that sent it.
_OK_MIN: Final = 200
_OK_MAX: Final = 300

#: What a phone is allowed to say. `at` is absent ON PURPOSE — the fleet stamps it
#: from its own clock so a beat cannot backdate itself, and `viaLan` is absent
#: because it is this relay's testimony, not the phone's.
_FROM_PHONE: Final = (
    "device",
    "app",
    "version",
    "startedAt",
    "streaming",
    "charging",
    "micOk",
)


class RelayRejected(Exception):
    """The body was not a beat this relay will pass on."""


def relayed(raw: bytes) -> dict[str, Any]:
    """The beat to forward: what the phone said, filtered, plus how it arrived.

    Raises `RelayRejected` rather than forwarding anything questionable — this is
    the boundary between an unauthenticated LAN caller and the fleet's store.
    """
    if len(raw) > MAX_BODY_BYTES:
        msg = f"body is {len(raw)} bytes"
        raise RelayRejected(msg)
    try:
        parsed = json.loads(raw)
    except (ValueError, UnicodeDecodeError) as exc:
        msg = f"not JSON: {exc}"
        raise RelayRejected(msg) from exc
    if not isinstance(parsed, dict):
        msg = f"not an object: {type(parsed).__name__}"
        raise RelayRejected(msg)

    device = parsed.get("device")
    if not isinstance(device, str) or not device:
        msg = "no device"
        raise RelayRejected(msg)
    if len(device) > MAX_DEVICE_LEN:
        msg = f"device name is {len(device)} chars"
        raise RelayRejected(msg)

    out: dict[str, Any] = {k: parsed[k] for k in _FROM_PHONE if k in parsed}
    # Stamped, never copied: see the module docstring.
    out["viaLan"] = True
    return out


def forward(beat: dict[str, Any], fleet_url: str = DEFAULT_FLEET_URL) -> bool:
    """POST one beat to the fleet. Returns whether it landed.

    Best-effort like every other part of this path: a relay that raised its own
    failures would be the tail wagging the dog. The phone learns by the status we
    return, and its own retry backoff (#886) decides what to do about it.
    """
    request = urllib.request.Request(
        f"{fleet_url}{BEAT_PATH}",
        data=json.dumps(beat).encode(),
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    try:
        with urllib.request.urlopen(request, timeout=FORWARD_TIMEOUT_S) as resp:
            status: int = resp.status
            return _OK_MIN <= status < _OK_MAX
    except (urllib.error.URLError, TimeoutError, OSError) as exc:
        _log.warning("could not forward a beat to %s: %s", fleet_url, exc)
        return False


class _Handler(BaseHTTPRequestHandler):
    fleet_url: str = DEFAULT_FLEET_URL

    # No @override: BaseHTTPRequestHandler declares no `do_POST` — it dispatches on
    # the method name at runtime, so this defines the handler rather than overriding.
    def do_POST(self) -> None:
        if self.path != BEAT_PATH:
            self.send_error(404)
            return
        length = int(self.headers.get("Content-Length") or 0)
        if length > MAX_BODY_BYTES:
            self.send_error(413)
            return
        try:
            beat = relayed(self.rfile.read(length))
        except RelayRejected as exc:
            _log.warning("refused a beat: %s", exc)
            self.send_error(400)
            return
        if not forward(beat, self.fleet_url):
            # 502, not 200: the phone must not read "the Mac took it" as "the fleet
            # knows", or a dead Isis would look like three healthy mics.
            self.send_error(502)
            return
        _log.info("relayed a beat from %s", beat.get("device"))
        self.send_response(204)
        self.end_headers()

    @override
    def log_message(self, format: str, *args: Any) -> None:
        """Quiet by default: the agent's log is for beats, not for one line per
        request from a phone that beats every hour anyway."""


def serve(port: int = DEFAULT_RELAY_PORT, fleet_url: str = DEFAULT_FLEET_URL) -> None:
    """Accept beats on the LAN and pass them to the fleet, forever."""
    handler = type("_BoundHandler", (_Handler,), {"fleet_url": fleet_url})
    server = ThreadingHTTPServer(("0.0.0.0", port), handler)
    _log.info("beat relay listening on :%d, forwarding to %s", port, fleet_url)
    server.serve_forever()

File: audiod/src/beat_relay.rs
//! LAN fallback for the mic heartbeat.
//!
//! Audio goes phone -> Mac over the LAN, but the beat goes phone -> Isis over
//! the VPN, so a phone at home with its tunnel off would record correctly and
//! still read `silent`, then `dead`. This relay lets the beat take the LAN too.
//!
//! ⚠ Not the ingest port, and not gated on the pause: `server::serve` closes
//! its listener while capture is paused, which is exactly when the heartbeat is
//! the only signal there is.
//!
//! Isis stays the single source of truth: this forwards and stores nothing, so
//! two places cannot disagree about which mics are alive.
//!
//! Hand-written HTTP, deliberately: one route, one unauthenticated LAN caller,
//! a handful of requests a month. `axum` and `tokio` are only dev-dependencies
//! here, and the capture daemon should not take them on for this.

use std::io::{BufRead, BufReader, Read, Write};
use std::net::{TcpListener, TcpStream};
use std::time::Duration;

/// The port the phone apps use for the fallback: the same port the fleet API
/// answers on, so the fallback is the identical request with `host` swapped for
/// `controlHost`.
pub const DEFAULT_RELAY_PORT: u16 = 8000;

/// Where a beat is posted, here and on the fleet.
pub const BEAT_PATH: &str = "/api/devices/heartbeat";
/// Bounded because the caller is unauthenticated: a beat is a few hundred bytes.
pub const MAX_BODY_BYTES: usize = 4096;
const MAX_DEVICE_LEN: usize = 64;
const MAX_HEAD_BYTES: usize = 8192;
const FORWARD_TIMEOUT: Duration = Duration::from_secs(8);
const READ_TIMEOUT: Duration = Duration::from_secs(10);

/// What a phone is allowed to say.
///
/// `at` is absent on purpose: the fleet stamps it, so a beat cannot backdate
/// itself. `viaLan` is the relay's testimony, not the phone's, and is added
/// below.
const FROM_PHONE: &[&str] = &[
    "device",
    "app",
    "version",
    "startedAt",
    "streaming",
    "charging",
    "micOk",
];

/// Why a body was not a beat this relay will pass on.
#[derive(Debug, PartialEq, Eq)]
pub struct Rejected(pub String);

/// The beat to forward: what the phone said, filtered, plus how it arrived.
/// An allowlist rather than a scrub, since the caller is unauthenticated: a key
/// not named here cannot reach Isis.
///
/// # Errors
/// If the body is oversized, is not a JSON object, or names no usable device.
pub fn relayed(raw: &[u8]) -> Result<serde_json::Value, Rejected> {
    if raw.len() > MAX_BODY_BYTES {
        return Err(Rejected(format!("body is {} bytes", raw.len())));
    }
    let parsed: serde_json::Value =
        serde_json::from_slice(raw).map_err(|e| Rejected(format!("not JSON: {e}")))?;
    let Some(obj) = parsed.as_object() else {
        return Err(Rejected("not an object".to_owned()));
    };
    let device = obj.get("device").and_then(serde_json::Value::as_str);
    match device {
        None | Some("") => return Err(Rejected("no device".to_owned())),
        Some(d) if d.chars().count() > MAX_DEVICE_LEN => {
            return Err(Rejected(format!(
                "device name is {} chars",
                d.chars().count()
            )));
        }
        Some(_) => {}
    }
    let mut out = serde_json::Map::new();
    for key in FROM_PHONE {
        if let Some(v) = obj.get(*key) {
            out.insert((*key).to_owned(), v.clone());
        }
    }
    // Stamped, never copied.
    out.insert("viaLan".to_owned(), serde_json::Value::Bool(true));
    Ok(serde_json::Value::Object(out))
}

/// POST one beat to the fleet. Returns whether it landed.
///
/// Best-effort: the phone learns from the status returned, and its own retry
/// backoff decides what to do.
pub fn forward(beat: &serde_json::Value, fleet_url: &str) -> bool {
    let url = format!("{}{BEAT_PATH}", fleet_url.trim_end_matches('/'));
    // Its own agent: `ureq`'s bare functions share one process-wide connection
    // pool, and a stale pooled socket is an unreproducible 502.
    let agent = ureq::AgentBuilder::new()
        .timeout(FORWARD_TIMEOUT)
        .max_idle_connections(0)
        .build();
    // `send_bytes` with the header set by hand: ureq is built without the
    // `json` feature that provides `send_json`.
    let body = beat.to_string();
    match agent
        .post(&url)
        .set("Content-Type", "application/json")
        .send_bytes(body.as_bytes())
    {
        Ok(_) => true,
        Err(ureq::Error::Status(status, _)) => {
            tracing::warn!(status, "beat-relay: the fleet refused a beat");
            false
        }
        Err(err) => {
            tracing::warn!(%err, url, "beat-relay: could not forward a beat");
            false
        }
    }
}

/// The parsed head of one HTTP request: enough for one route and no more.
#[derive(Debug, PartialEq, Eq)]
pub struct Head {
    pub method: String,
    pub path: String,
    pub content_length: usize,
}

/// Read the request line and headers, bounded.
///
/// Bounded on both line count and byte count, so an unauthenticated caller
/// sending headers forever costs nothing.
///
/// # Errors
/// If the stream ends, the head is oversized, or the request line is not
/// `METHOD PATH VERSION`.
pub fn read_head<R: BufRead>(reader: &mut R) -> Result<Head, Rejected> {
    let mut line = String::new();
    let mut taken = 0usize;
    reader
        .read_line(&mut line)
        .map_err(|e| Rejected(format!("no request line: {e}")))?;
    taken += line.len();
    let mut parts = line.split_whitespace();
    let (Some(method), Some(path)) = (parts.next(), parts.next()) else {
        return Err(Rejected("malformed request line".to_owned()));
    };
    let (method, path) = (method.to_owned(), path.to_owned());
    let mut content_length = 0usize;
    loop {
        let mut header = String::new();
        let n = reader
            .read_line(&mut header)
            .map_err(|e| Rejected(format!("bad header: {e}")))?;
        if n == 0 {
            return Err(Rejected("head ended without a blank line".to_owned()));
        }
        taken += n;
        if taken > MAX_HEAD_BYTES {
            return Err(Rejected(format!("head is over {MAX_HEAD_BYTES} bytes")));
        }
        let header = header.trim_end();
        if header.is_empty() {
            break;
        }
        if let Some((name, value)) = header.split_once(':')
            && name.eq_ignore_ascii_case("content-length")
        {
            content_length = value
                .trim()
                .parse()
                .map_err(|_| Rejected("unparseable Content-Length".to_owned()))?;
        }
    }
    Ok(Head {
        method,
        path,
        content_length,
    })
}

/// Answer with an empty body. Success is 204 (the relay has nothing to
/// return); a failed forward is 502, never 204, so "the Mac took it" never
/// reads as "the fleet knows".
fn respond(stream: &mut TcpStream, status: u16, reason: &str) {
    let _ = write!(
        stream,
        "HTTP/1.1 {status} {reason}\r\nContent-Length: 0\r\nConnection: close\r\n\r\n"
    );
    let _ = stream.flush();
}

fn handle(stream: &mut TcpStream, fleet_url: &str) {
    let _ = stream.set_read_timeout(Some(READ_TIMEOUT));
    let _ = stream.set_write_timeout(Some(READ_TIMEOUT));
    let peer = stream
        .peer_addr()
        .map_or_else(|_| "?".to_owned(), |a| a.to_string());
    let mut reader = BufReader::new(match stream.try_clone() {
        Ok(s) => s,
        Err(err) => {
            tracing::warn!(%err, "beat-relay: cannot read the connection");
            return;
        }
    });
    let head = match read_head(&mut reader) {
        Ok(head) => head,
        Err(Rejected(why)) => {
            tracing::warn!(why, peer, "beat-relay: bad request");
            respond(stream, 400, "Bad Request");
            return;
        }
    };
    if head.method != "POST" || head.path != BEAT_PATH {
        respond(stream, 404, "Not Found");
        return;
    }
    if head.content_length > MAX_BODY_BYTES {
        respond(stream, 413, "Payload Too Large");
        return;
    }
    let mut body = vec![0u8; head.content_length];
    if reader.read_exact(&mut body).is_err() {
        respond(stream, 400, "Bad Request");
        return;
    }
    let beat = match relayed(&body) {
        Ok(beat) => beat,
        Err(Rejected(why)) => {
            tracing::warn!(why, peer, "beat-relay: refused a beat");
            respond(stream, 400, "Bad Request");
            return;
        }
    };
    if !forward(&beat, fleet_url) {
        respond(stream, 502, "Bad Gateway");
        return;
    }
    let device = beat.get("device").and_then(serde_json::Value::as_str);
    tracing::info!(device, "beat-relay: relayed a beat");
    respond(stream, 204, "No Content");
}

/// Accept beats on the LAN and pass them to the fleet, forever.
///
/// One thread per connection: a phone whose socket stalls must not hold up the
/// next phone's beat, and the volume is too low to justify anything cleverer.
pub fn serve(port: u16, fleet_url: &str) -> std::io::Error {
    let listener = match TcpListener::bind(("0.0.0.0", port)) {
        Ok(l) => l,
        Err(err) => return err,
    };
    tracing::info!(port, fleet_url, "beat-relay: listening");
    for stream in listener.incoming() {
        match stream {
            Ok(mut stream) => {
                let fleet = fleet_url.to_owned();
                std::thread::spawn(move || handle(&mut stream, &fleet));
            }
            Err(err) => tracing::warn!(%err, "beat-relay: accept failed"),
        }
    }
    std::io::Error::other("the accept loop ended")
}
